### ISSUE TITLE
TensorFlow Lite conversion fails due to tf.linalg.diag_part

### ISSUE THREAD

**Issue by ingo-m** [tensorflow/tensorflow#45321]
**System information**
- OS Platform and Distribution: google colab (same issue locally on Debian GNU/Linux 10)
- TensorFlow installed from: PyPI (`pip install`)
- TensorFlow version: 2.3.0 (google colab), and 2.3.1 (locally)

**Command used to run the converter or code if you’re using the Python API**
I am trying to convert a model that uses [tf.linalg.diag_part](https://www.tensorflow.org/api_docs/python/tf/linalg/diag_part) to TensorFlow Lite. As far as I understand, `tf.linalg.diag_part` is not directly supported, but it should be possible to convert the model by setting `converter.target_spec.supported_ops` (as described [here](https://www.tensorflow.org/lite/guide/ops_select#convert_a_model). `tf.linalg.diag_part` seems to rely on `MatrixDiagPartV3`, which is included in the [list](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/flex/allowlisted_flex_ops.cc) of operations that can be enabled. Here is a minimal example ([colab notebook](https://colab.research.google.com/drive/1-sYbh6In8Dubki0UtcnzjmEZuMiMa-Gd?usp=sharing)):
```python
import tensorflow as tf

model_input = tf.keras.Input(shape=(3, 3), batch_size=1)

diagonal = tf.linalg.diag_part(model_input)

model = tf.keras.models.Model(inputs=model_input, outputs=diagonal)

converter = tf.lite.TFLiteConverter.from_keras_model(model)

# Enable TensorFlow ops that are not directly supported by tf lite
# https://www.tensorflow.org/lite/guide/ops_select#convert_a_model
converter.target_spec.supported_ops = [
    tf.lite.OpsSet.TFLITE_BUILTINS,
    tf.lite.OpsSet.SELECT_TF_OPS,
    ]

tflite_model = converter.convert()
```

**The output from the converter invocation**

```bash
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/training/tracking/tracking.py:111: Model.state_updates (from tensorflow.python.keras.engine.training) is deprecated and will be removed in a future version.
Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/training/tracking/tracking.py:111: Layer.updates (from tensorflow.python.keras.engine.base_layer) is deprecated and will be removed in a future version.
Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.
INFO:tensorflow:Assets written to: /tmp/tmpkg1mi0e5/assets

---------------------------------------------------------------------------

Exception                                 Traceback (most recent call last)

/usr/local/lib/python3.6/dist-packages/tensorflow/lite/python/convert.py in toco_convert_protos(model_flags_str, toco_flags_str, input_data_str, debug_info_str, enable_mlir_converter)
    198                                                  debug_info_str,
--> 199                                                  enable_mlir_converter)
    200       return model_str

5 frames

Exception: /usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/def_function.py:1073:0: error: 'tf.MatrixDiagPartV3' op is neither a custom op nor a flex op
/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/def_function.py:1167:0: note: called from
/usr/local/lib/python3.6/dist-packages/tensorflow/lite/python/lite.py:804:0: note: called from
<ipython-input-1-9932e84c0ae3>:23:0: note: called from
/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2882:0: note: called from
/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2822:0: note: called from
/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718:0: note: called from
/usr/local/lib/python3.6/dist-packages/ipykernel/zmqshell.py:537:0: note: called from
/usr/local/lib/python3.6/dist-packages/ipykernel/ipkernel.py:208:0: note: called from
/usr/local/lib/python3.6/dist-packages/ipykernel/kernelbase.py:399:0: note: called from
/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/def_function.py:1073:0: note: see current operation: %0 = "tf.MatrixDiagPartV3"(%arg0, %cst, %cst_0) {T = f32, _cloned = true, align = "RIGHT_LEFT", device = ""} : (tensor<1x3x3xf32>, tensor<i32>, tensor<f32>) -> tensor<1x3xf32>
<unknown>:0: error: failed while converting: 'main': Ops that need custom implementation (enabled via setting the -emit-custom-ops flag):
	tf.MatrixDiagPartV3 {T = f32, _cloned = true, align = "RIGHT_LEFT", device = ""}
<unknown>:0: note: see current operation: "func"() ( {
^bb0(%arg0: tensor<1x3x3xf32>):  // no predecessors
  %cst = "std.constant"() {value = dense<0> : tensor<i32>} : () -> tensor<i32>
  %cst_0 = "std.constant"() {value = dense<0.000000e+00> : tensor<f32>} : () -> tensor<f32>
  %0 = "tf.MatrixDiagPartV3"(%arg0, %cst, %cst_0) {T = f32, _cloned = true, align = "RIGHT_LEFT", device = ""} : (tensor<1x3x3xf32>, tensor<i32>, tensor<f32>) -> tensor<1x3xf32>
  "std.return"(%0) : (tensor<1x3xf32>) -> ()
}) {sym_name = "main", tf.entry_function = {control_outputs = "", inputs = "input_1", outputs = "Identity"}, type = (tensor<1x3x3xf32>) -> tensor<1x3xf32>} : () -> ()


During handling of the above exception, another exception occurred:

ConverterError                            Traceback (most recent call last)

/usr/local/lib/python3.6/dist-packages/tensorflow/lite/python/convert.py in toco_convert_protos(model_flags_str, toco_flags_str, input_data_str, debug_info_str, enable_mlir_converter)
    200       return model_str
    201     except Exception as e:
--> 202       raise ConverterError(str(e))
    203 
    204   if distutils.spawn.find_executable(_toco_from_proto_bin) is None:

ConverterError: /usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/def_function.py:1073:0: error: 'tf.MatrixDiagPartV3' op is neither a custom op nor a flex op
/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/def_function.py:1167:0: note: called from
/usr/local/lib/python3.6/dist-packages/tensorflow/lite/python/lite.py:804:0: note: called from
<ipython-input-1-9932e84c0ae3>:23:0: note: called from
/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2882:0: note: called from
/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2822:0: note: called from
/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718:0: note: called from
/usr/local/lib/python3.6/dist-packages/ipykernel/zmqshell.py:537:0: note: called from
/usr/local/lib/python3.6/dist-packages/ipykernel/ipkernel.py:208:0: note: called from
/usr/local/lib/python3.6/dist-packages/ipykernel/kernelbase.py:399:0: note: called from
/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/def_function.py:1073:0: note: see current operation: %0 = "tf.MatrixDiagPartV3"(%arg0, %cst, %cst_0) {T = f32, _cloned = true, align = "RIGHT_LEFT", device = ""} : (tensor<1x3x3xf32>, tensor<i32>, tensor<f32>) -> tensor<1x3xf32>
<unknown>:0: error: failed while converting: 'main': Ops that need custom implementation (enabled via setting the -emit-custom-ops flag):
	tf.MatrixDiagPartV3 {T = f32, _cloned = true, align = "RIGHT_LEFT", device = ""}
<unknown>:0: note: see current operation: "func"() ( {
^bb0(%arg0: tensor<1x3x3xf32>):  // no predecessors
  %cst = "std.constant"() {value = dense<0> : tensor<i32>} : () -> tensor<i32>
  %cst_0 = "std.constant"() {value = dense<0.000000e+00> : tensor<f32>} : () -> tensor<f32>
  %0 = "tf.MatrixDiagPartV3"(%arg0, %cst, %cst_0) {T = f32, _cloned = true, align = "RIGHT_LEFT", device = ""} : (tensor<1x3x3xf32>, tensor<i32>, tensor<f32>) -> tensor<1x3xf32>
  "std.return"(%0) : (tensor<1x3xf32>) -> ()
}) {sym_name = "main", tf.entry_function = {control_outputs = "", inputs = "input_1", outputs = "Identity"}, type = (tensor<1x3x3xf32>) -> tensor<1x3xf32>} : () -> ()
```

**Also, please include a link to the saved model or GraphDef**

The minimal example does not include a saved model, but see here for the [colab notebook](https://colab.research.google.com/drive/1-sYbh6In8Dubki0UtcnzjmEZuMiMa-Gd?usp=sharing).

**Failure details**
n/a

**RNN conversion support**
n/a

**Any other info / logs**

I encountered this conversion problem both in TensorFlow 2.3.0 (google colab), and in 2.3.1 (locally, installed via `pip`). Taking the diagonal of matrix is an operation that can be relevant when trying to build lightweight models, e.g. for use in resource constrained environments. In my use case, I am trying to take the diagonal of a Gram matrix in an image style application.

**Comment by ravikyram** [comment]
@ingo-m 

After adding `converter.allow_custom_ops=True` I am not seeing any issue. Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/cecb0727a4ee880c07db84cb1fdcdd14/untitled555.ipynb).Thanks!

**Comment by ingo-m** [comment]
@ravikyram Thanks for your quick reply. Nice, I wasn't aware of that. Perhaps the missing line should be added to the [guide](https://www.tensorflow.org/lite/guide/ops_select#convert_a_model), so that others don't run into the same problem? I submitted a pull request: https://github.com/tensorflow/tensorflow/pull/45356

**Comment by ravikyram** [comment]
@ingo-m 

Please,close this thread if your issue was resolved. Thanks!

**Comment by ingo-m** [comment]
Unfortunately I have to reopen this issue - I had overlooked that when enabling custom operations (`converter.allow_custom_ops = True`), the model can be compiled and saved, but can't perform inference. I updated the [colab notebook](https://colab.research.google.com/drive/1-sYbh6In8Dubki0UtcnzjmEZuMiMa-Gd?usp=sharing) with the complete example, tested in TensorFlow 2.3.1.

Specifically, after loading the TensorFlow lite model from disk, `interpreter.allocate_tensors()` fails with the following traceback:
```bash
---------------------------------------------------------------------------

RuntimeError                              Traceback (most recent call last)

<ipython-input-4-e8f284b41149> in <module>()
----> 1 interpreter.allocate_tensors()

/usr/local/lib/python3.6/dist-packages/tensorflow/lite/python/interpreter.py in allocate_tensors(self)
    241   def allocate_tensors(self):
    242     self._ensure_safe()
--> 243     return self._interpreter.AllocateTensors()
    244 
    245   def _safe_to_run(self):

RuntimeError: Encountered unresolved custom op: MatrixDiagPartV3.Node number 0 (MatrixDiagPartV3) failed to prepare.

```

**Comment by ingo-m** [comment]
PS: The user [guide](https://www.tensorflow.org/lite/guide/ops_select#python) mentions: "TensorFlow Lite with select TensorFlow ops are available in the TensorFlow pip package version since 2.3 for Linux", so the [colab example](https://colab.research.google.com/drive/1-sYbh6In8Dubki0UtcnzjmEZuMiMa-Gd?usp=sharing) should work, right?

**Comment by ymodak** [comment]
Apologies for the delay in response. This is fixed latest TF 2.4.1 release.
See [gist](https://colab.research.google.com/gist/ymodak/c11224c3f457417dd72b3c246c9fd928/tflite_diag_part_min_example.ipynb). Thanks!

**Comment by oawxkw** [comment]
I also observed the following API aliases or similar APIs can cause the same issue in older versions of tensorflow.
Users should be cautious when using them on CPU up to tensorflow 2.3.4 (v2.3.3-137-gea90cf44f73).

- `tf.linalg.diag_part`, `tf.compat.v1.linalg.diag_part`
- `tf.linalg.tensor_diag`, `tf.compat.v1.linalg.tensor_diag`, `tf.compat.v1.diag`
- `tf.compat.v1.matrix_diag_part`

It seems to be fixed in tensorflow 2.4.0 (v2.4.0-rc4-71-g582c8d236cb) and later versions.
